### DISCUSSION TITLE
Using StringFormat when binding from code // "Expected end of expression."

### DISCUSSION THREAD

**Discussion by NoizeDaemon** [AvaloniaUI/Avalonia#13171] [Q&A] [answered]
Hi!

I'm trying to display a list of floats (with varying counts) on a class as columns in a `DataGrid`.
To achieve this, I remove the old columns from the grid and add new ones from code behind when the `DataContext` of the grid changes.

Here is how I define the cell template and bind the value:
```c#
  var relativeFrequencyPath = $"{nameof(ResultLine.RelativeFrequencies)}[{i}];
  Debug.WriteLine(relativeFrequencyPath);

  var cellTemplate = new FuncDataTemplate<ResultLine>((value, namescope) =>
      new TextBlock()
      {
          [!TextBlock.TextProperty] = new Binding(relativeFrequencyPath)
      });

  var newColumn = new DataGridTemplateColumn()
  {
      Header = header,
      CellTemplate = cellTemplate
  };
```

That works fine:
![image](https://github.com/AvaloniaUI/Avalonia/assets/42150677/d356d5b7-e4c5-4011-91e1-5cdc341443d8)
![image](https://github.com/AvaloniaUI/Avalonia/assets/42150677/0f35d407-f08e-420a-83eb-f599b9d8fd3d)

However, once I add the `StringFormat` like so:
```c#
  var relativeFrequencyPath = $"{nameof(ResultLine.RelativeFrequencies)}[{i}], StringFormat=P1";
```
I get this error:
```
Exception thrown: 'Avalonia.Data.Core.ExpressionParseException' in Avalonia.Markup.dll
An unhandled exception of type 'Avalonia.Data.Core.ExpressionParseException' occurred in Avalonia.Markup.dll
Expected end of expression.
```

### Am I using `StringFormat` wrong? Or does it not work when binding from code?

---

If I use a single `DataGridTemplateColumn` in XAML and just display all the values in an `ItemsRepeater`, it works properly:
```axaml
	<DataGridTemplateColumn.CellTemplate>
		<DataTemplate DataType="models:ResultLine">
			<ItemsRepeater ItemsSource="{Binding RelativeFrequencies}">
				<ItemsRepeater.Layout>
					<StackLayout Orientation="Horizontal" Spacing="20" />
				</ItemsRepeater.Layout>
				<ItemsRepeater.ItemTemplate>
					<DataTemplate>
						<TextBlock Width="200" Text="{Binding, StringFormat=P1}" />
					</DataTemplate>
				</ItemsRepeater.ItemTemplate>
			</ItemsRepeater>
		</DataTemplate>
	</DataGridTemplateColumn.CellTemplate>
```

---

A random curious thing is that it seems to run fine the first few times. This is my output:
```
RelativeFrequencies[0], StringFormat=P1
RelativeFrequencies[1], StringFormat=P1
RelativeFrequencies[2], StringFormat=P1
RelativeFrequencies[3], StringFormat=P1
Exception thrown: [...]
```

Altough that could also just mean it finishes my loop before evaluating the binding path 🤷‍♀️

---

I'd appreciate any help!

**Comment by maxkatz6** [comment, marked as answer]
> However, once I add the StringFormat like so:

StringFormat is not part of Path property. It's a separate property. You can set it on Binding object as well.

**Reply by NoizeDaemon** [reply]
Thank you!

It was that simple 🤦‍♀️😅
```c#
[!TextBlock.TextProperty] = new Binding(relativeFrequencyPath) { StringFormat = "P1"}
```

For some reason I'm only ever looking at constructor overloads & never at properties on the class 🙃
